### PR TITLE
[7.16] Add hidden data streams for endpoint actions and responses

### PR DESCRIPTION
**Pull request by ashokaditya**
custom mapping for hidden data streams named `endpoint-actions` and `endpoint-action-responses`.

`.endpoint-actions` mapping follows 
```json5
{
  "mappings": {
    "dynamic": "false",
    "_meta": {
        "version": "8.0.0"
    },
    "_data_stream_timestamp": {
        "enabled": true
    },
    "properties": {
        "@timestamp": {
            "type": "date"
        },
        "action_id": {
            "type": "keyword"
        },
        "agents": {
            "type": "keyword"
        },
        "data": {
            "properties" : {
              "command": { "type": "keyword" },
              "comment": { "type": "text" },
              "file": {
                  "properties": {
                      "blob": { "type": "binary" }
                    }
                }
            }
        },
        "expiration": {
            "type": "date"
        },
        "input_type": {
            "type": "keyword"
        },
        "type": {
            "type": "keyword"
        },
        "user_id": {
            "type": "keyword"
        }
    }
  }
}
```

and `.endpoint-action-responses` follows
```json5
{
  "mappings": {
    "dynamic": "false",
    "_meta": {
        "version": "8.0.0"
    },
    "_data_stream_timestamp": {
        "enabled": true
    },
    "properties": {
        "@timestamp": {
            "type": "date"
        },
        "action_id": {
            "type": "keyword"
        },
        "agent_id": {
            "type": "keyword"
        },
        "completed_at": {
            "type": "date"
        },
        "data": {
            "properties": { 
            "command": { "type": "keyword" },
            "comment": { "type": "text" },
            "file": {
            "properties": {
                "blob": { "type": "binary" },
                "path": { "type": "text" },
                "execute": {
                  "properties":{
                      "status": { "type": "keyword" },
                      "error": {
                      "type": "text",
                        "fields":{
                            "keyword" : {
                              "type" : "keyword",
                              "ignore_above": 1024
                          }
                        }
                      }
                  }
                }
            }
            }  
        }
        },
        "status": {
          "type": "keyword"
        },
        "error": {
            "type": "text",
            "fields": {
                "keyword": {
                    "type": "keyword",
                    "ignore_above": 1024
                }
            }
        },
        "started_at": {
            "type": "date"
        }
    }
  }
}
```

After creating the schemas for the hidden data streams one can add documents to the hidden indices like so.

**For action requests:**
```json5
POST .logs-endpoint.actions-01/_doc
{
    "action_id": "9383a2f3-4347-4f23-9331-ae05a3bc42bf",
		"@timestamp": "2021-09-10T13:24:54.005Z",
		"expiration": "2021-09-10T13:24:54.005Z",
		"type": "INPUT_ACTION",
		"input_type": "endpoint",
		"agent": {
		  "id": [
			  "fd041791-f467-474d-bb70-5dab47004483", "db19b5b5-59b0-4a22-aebb-e96895d8fd0c"
			  ]
			},
			"user_id": "elastic",
			"data": {
			  "command": "isolate",
			  "comment": "new index doc"
		  } 
}
```

**For action responses**
```json5
POST .logs-endpoint.action.responses-01/_doc
{
  "@timestamp": "2021-09-12T13:24:29.193Z",
  "data": {
    "command": "isolate", 
    "comment": "new responses index"
  },
  "agent": {
      "id":  "fd041791-f467-474d-bb70-5dab47004483"
  },
  "action_id": "9383a2f3-4347-4f23-9331-ae05a3bc42bf",
  "agent_id": "db19b5b5-59b0-4a22-aebb-e96895d8fd0c",
  "completed_at": "2021-09-11T13:24:29.193Z",
  "started_at": "2021-09-11T13:24:29.194Z"
}

```

Requesting the mappings for each of these indices by running

```json5
GET .logs-endpoint.action*/_mapping
``` 

created above gives us:

```json5
{
  ".ds-.logs-endpoint.actions-01-2021.09.16-000001" : {
    "mappings" : {
      "dynamic" : "false",
      "_meta" : {
        "package" : {
          "name" : "endpoint"
        },
        "managed_by" : "ingest-manager",
        "managed" : true
      },
      "_data_stream_timestamp" : {
        "enabled" : true
      },
      "dynamic_templates" : [
        {
          "strings_as_keyword" : {
            "match_mapping_type" : "string",
            "mapping" : {
              "ignore_above" : 1024,
              "type" : "keyword"
            }
          }
        }
      ],
      "date_detection" : false,
      "properties" : {
        "@timestamp" : {
          "type" : "date"
        },
        "action_id" : {
          "type" : "keyword",
          "ignore_above" : 1024
        },
        "agent" : {
          "properties" : {
            "id" : {
              "type" : "keyword",
              "ignore_above" : 1024
            }
          }
        },
        "agents" : {
          "type" : "alias",
          "path" : "agent.id"
        },
        "data" : {
          "properties" : {
            "command" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "comment" : {
              "type" : "text"
            }
          }
        },
        "data_stream" : {
          "properties" : {
            "dataset" : {
              "type" : "constant_keyword"
            },
            "namespace" : {
              "type" : "constant_keyword"
            },
            "type" : {
              "type" : "constant_keyword"
            }
          }
        },
        "ecs" : {
          "properties" : {
            "version" : {
              "type" : "keyword",
              "ignore_above" : 1024
            }
          }
        },
        "event" : {
          "properties" : {
            "action" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "agent_id_status" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "category" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "created" : {
              "type" : "date"
            },
            "end" : {
              "type" : "date"
            },
            "hash" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "id" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "ingested" : {
              "type" : "date"
            },
            "outcome" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "start" : {
              "type" : "date"
            },
            "type" : {
              "type" : "keyword",
              "ignore_above" : 1024
            }
          }
        },
        "expiration" : {
          "type" : "date"
        },
        "input_type" : {
          "type" : "keyword",
          "ignore_above" : 1024
        },
        "type" : {
          "type" : "keyword",
          "ignore_above" : 1024
        },
        "user_id" : {
          "type" : "keyword",
          "ignore_above" : 1024
        }
      }
    }
  },
  ".ds-.logs-endpoint.action.responses-01-2021.09.16-000001" : {
    "mappings" : {
      "dynamic" : "false",
      "_meta" : {
        "package" : {
          "name" : "endpoint"
        },
        "managed_by" : "ingest-manager",
        "managed" : true
      },
      "_data_stream_timestamp" : {
        "enabled" : true
      },
      "dynamic_templates" : [
        {
          "strings_as_keyword" : {
            "match_mapping_type" : "string",
            "mapping" : {
              "ignore_above" : 1024,
              "type" : "keyword"
            }
          }
        }
      ],
      "date_detection" : false,
      "properties" : {
        "@timestamp" : {
          "type" : "date"
        },
        "action_id" : {
          "type" : "keyword",
          "ignore_above" : 1024
        },
        "agent" : {
          "properties" : {
            "id" : {
              "type" : "keyword",
              "ignore_above" : 1024
            }
          }
        },
        "agent_id" : {
          "type" : "alias",
          "path" : "agent.id"
        },
        "completed_at" : {
          "type" : "date"
        },
        "data" : {
          "properties" : {
            "command" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "comment" : {
              "type" : "text"
            }
          }
        },
        "data_stream" : {
          "properties" : {
            "dataset" : {
              "type" : "constant_keyword"
            },
            "namespace" : {
              "type" : "constant_keyword"
            },
            "type" : {
              "type" : "constant_keyword"
            }
          }
        },
        "ecs" : {
          "properties" : {
            "version" : {
              "type" : "keyword",
              "ignore_above" : 1024
            }
          }
        },
        "event" : {
          "properties" : {
            "action" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "agent_id_status" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "category" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "created" : {
              "type" : "date"
            },
            "end" : {
              "type" : "date"
            },
            "hash" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "id" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "ingested" : {
              "type" : "date"
            },
            "outcome" : {
              "type" : "keyword",
              "ignore_above" : 1024
            },
            "start" : {
              "type" : "date"
            },
            "type" : {
              "type" : "keyword",
              "ignore_above" : 1024
            }
          }
        },
        "started_at" : {
          "type" : "date"
        },
        "status" : {
          "type" : "keyword",
          "ignore_above" : 1024
        }
      }
    }
  }
}

```

### CODE REVIEW

File: custom_subsets/elastic_endpoint/action_responses/action_responses.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  base:
+    fields:
+      "@timestamp": {}
+      action_id: {}
+      completed_at: {}
+      data:
+        fields: "*"
+      status: {}
+      started_at: {}
+  agent:
```
Comment (ashokaditya) [inline custom_subsets/elastic_endpoint/action_responses/action_responses.yaml:13]: Should this be on this level as is `base` or should `agent` be on the same level as `@timestamp` like? 🤔 

```json5
---
name: action_responses
fields:
  base:
    fields:
      "@timestamp": {}
      action_id: {}
      completed_at: {}
      data:
        fields: "*"
      status: {}
      started_at: {}
      agent:
        fields:
          id: {}
...
...
```

@pzl

Comment (pzl) [reply custom_subsets/elastic_endpoint/action_responses/action_responses.yaml:13]: I do not believe it should be a `base` field. `agent` is its own group, I believe. Certainly, how it is follows the pattern set in the other subset files

File: custom_subsets/elastic_endpoint/actions/actions.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+name: actions
+fields:
+  base:
+    fields:
+      "@timestamp": {}
+      action_id: {}
+      data:
+        fields: "*"
+      expiration: {}
+      input_type: {}
```
Comment (ashokaditya) [inline custom_subsets/elastic_endpoint/actions/actions.yaml:11]: I don't think we need to have `input_type` anymore, now that we've only endpoint data here. What about `type` do we still need that? @pzl

Comment (pzl) [reply custom_subsets/elastic_endpoint/actions/actions.yaml:11]: we don't need them for our purposes, no.
We left that to be a close copy, if not perfectly identical, to the fleet action

**Review by pzl — APPROVED** [review]
lgtm,

we can probably revisit the `event.*` fields before we cut a release. We populate `event.ingested` from the pipeline, but I wonder if any of the other `event.*` fields get populated with anything. Same with `data_stream`. Curious if that gets automatically populated

File: custom_schemas/custom_action.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
+---
```
Comment (jonathan-buttner) [inline custom_schemas/custom_action.yml:1]: Do we need this file anymore if everything is under `base` now?

File: custom_schemas/custom_action_responses.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+---
```
Comment (jonathan-buttner) [inline custom_schemas/custom_action_responses.yml:1]: Same with this file, I don't think we need it since everything is under `base`.

**Review by jonathan-buttner — COMMENTED** [review]
@ashokaditya could you reach out to the ECS folks and see if there's a way to get fields at the root level of a document without needing to place them under `base`. If base is the only place they can go, we should ask them for a feature request to allow custom schema to be placed at the root of the document if we plan on adding a lot of fields under base in the future.

Also could we ask if there's a way to mark a schema as reusable on top of base?

All of this is mainly to avoid having to place tons of fields inside of the base extension file.

Another thing you could try is to copy `custom_base.yml` to a new file and see if you have multiple files that extend base.

File: custom_schemas/custom_action_responses.yml (hunk 4 above)
Comment (ashokaditya) [reply custom_schemas/custom_action_responses.yml]: done d5010cb9c4629413305939ff05e4ba550e5c75fd

File: custom_schemas/custom_action.yml (hunk 3 above)
Comment (ashokaditya) [reply custom_schemas/custom_action.yml]: done d5010cb9c4629413305939ff05e4ba550e5c75fd

**Review by jonathan-buttner — APPROVED** [review]

**Review by joeypoon — APPROVED** [review]

**Pull request merged**